CORE-902 content portal routing

TylerZeroMaster opened the pull request:
https://openstax.atlassian.net/browse/CORE-902


Note: The cookieyes button does not seem to appear when running locally (I suspect it's added by javascript that doesn't run locally, but I am not 100% sure).

I needed to change some stuff about content routing in order for the portal to stay alive through content links. Maybe there's a better way, but I added `portalName` to `Params` and have it propagate through `ContentLink`. Here's a more detailed explanation:

1. `ContentLink` creates a url using `ContentRoute` `getUrl` inside `getBookPageUrlAndParams` [here](https://github.com/openstax/rex-web/blob/9294afd1a0f92e6ec3e0841a18f6acf7c2b143e1/src/app/content/components/ContentLink.tsx#L65).
2. The above url returns a non-portal url if the portal name is not included in `Params` (because `getUrl` returns urls based on matching params).
3. Content link href is transformed to a relative path [here](https://github.com/openstax/rex-web/blob/9294afd1a0f92e6ec3e0841a18f6acf7c2b143e1/src/app/content/components/ContentLink.tsx#L67).
4. The above function uses common parts of the url to perform the transformation. If `getBookPageUrlAndParams` returns a url that does not have a portal, then the relative url that is ultimately used in the content link will not have a portal either.

I did not make this change at a deeper level because having portal name in params elsewhere would interfere with things like fetching/receiving content.

Aside: I am not a huge fan of using regex for the navigation selector. I wondered if there was a way to use the content routes to extract the portal name instead, or if `portalName` should be added to the redux state. It seemed pretty closely linked to the pathname, which is why I decided to do it this way. Any suggestions would be welcome.

## Desktop
<img width="985" alt="Screenshot 2025-05-14 at 10 29 45 AM" src="https://github.com/user-attachments/assets/839b7875-5b10-461c-9d08-54bf02ba4463" />

## Mobile
<img width="377" alt="Screenshot 2025-05-14 at 10 29 59 AM" src="https://github.com/user-attachments/assets/59495518-9b06-4616-8cbd-15c10f62fe58" />

## Contact Dialog
<img width="1396" alt="Screenshot 2025-05-14 at 10 30 18 AM" src="https://github.com/user-attachments/assets/07be9689-eb48-4019-9558-e0600fe7d130" />

## Content Links
<img width="632" alt="Screenshot 2025-05-14 at 10 32 35 AM" src="https://github.com/user-attachments/assets/72ce8566-37bf-4a34-9324-3d44ae66e2d1" />

TylerZeroMaster committed:
Add unique behavior for portal

TylerZeroMaster committed:
Update content links for portal pages

This seemed like the least roundabout way to do it

TylerZeroMaster committed:
Update snapshots

Not sure why these changed... they look exactly the same

TylerZeroMaster committed:
Lint

TylerZeroMaster committed:
Use previously filtered handlers in footer test

staxly[bot] committed:
Merge branch 'main' into core-902-content-portal-routing

RoyEJohnson reviewed (approved):
I appreciate the description of what you were doing.
I left some comments for your consideration, but it looks fine to me.

TomWoodward commented on src/app/navigation/selectors.ts:
```diff
@@ -14,6 +14,11 @@ export const pathname = createSelector(
 
+export const portalName = createSelector(
+  pathname,
+  (state) => state.match(/\/portal\/([^/]+)\//)?.[1]
```
`navigation.match.params.portalName` will have the portal name if there is one

TomWoodward reviewed (changes requested):
looks good to me

TylerZeroMaster committed:
Wrap ContactDialog in callback

TylerZeroMaster committed:
Remove iframe border

TylerZeroMaster committed:
Use content.params.portalName instead of regex

TylerZeroMaster committed:
Add missing license link

TylerZeroMaster committed:
Update tests to use content params to set portalName

TylerZeroMaster committed:
Rename return value to avoid shadowing

staxly[bot] committed:
Merge branch 'main' into core-902-content-portal-routing

staxly[bot] committed:
Merge branch 'main' into core-902-content-portal-routing

staxly[bot] committed:
Merge branch 'main' into core-902-content-portal-routing

TylerZeroMaster committed:
Use match.params in navigation selectors

TylerZeroMaster committed:
Use guard to minimize references to `portalName`

TylerZeroMaster committed:
Make `getBookPageUrlAndParams` more generic

TylerZeroMaster commented on src/app/components/Footer/index.tsx:
```diff
@@ -293,7 +303,7 @@ const PortalColumn2 = () => {
         <FooterLinkMessage href='/privacy-policy' id='i18n:footer:column3:privacy-policy' />
       </LinkList>
-      <ContactDialog contactFormParams={contactFormParams} />
+      <Dialog contactFormParams={contactFormParams} />
```
Okay, but this pattern is used all over the place in os-webview. I literally copied this hook from os-webview. I didn't see this comment before. I will make this change and update the tests.

RoyEJohnson replied:
That's true; I was learning React as I ported osweb to use it, so there are some bad patterns in the codebase.

TylerZeroMaster committed:
Do not return contact dialog from hook.

TomWoodward commented on src/app/components/Footer/index.tsx:
```diff
@@ -209,4 +213,148 @@ const Footer = ({
+
+  React.useEffect(
+    () => {
+      if (typeof window === 'undefined' || !window.parent || window.parent === window) {
```
why are you checking `!window.parent || window.parent === window` here? this looks like a check to see if the rex page is embedded, thats not what we want right?

TomWoodward commented on src/app/content/components/ContentLink.tsx:
```diff
@@ -31,6 +31,7 @@ interface Props {
   handleClick?: () => void; // this one gets called instead of navigation
   navigate: typeof push;
   currentPath: string;
+  navigationParams: unknown;
```
why unknown?

TomWoodward commented on src/app/guards.ts:
```diff
@@ -47,3 +47,7 @@ export const isHtmlElementWithHighlight = (thing: unknown): thing is dom.HTMLEle
   typeof thing === 'string' && thing in obj;
+
+export const isPortaled = (thing: unknown): thing is {portalName: string} => (
+  typeof thing === 'object' && thing != null && 'portalName' in thing
```
technically this'll still match `{portalName: undefined}` or `{portalName: ''}`, you might want to throw a `&& thing.portalName` on the end. but also wouldn't it be just as easy to check `params?.portalName` inline?

TylerZeroMaster committed:
Improve type checking and usage for navigation params portalName

TylerZeroMaster committed:
Stop assuming portaled footer will be embedded

handle window being undefined

Fix event listener mock referencing unmounted component

staxly[bot] committed:
Merge branch 'main' into core-902-content-portal-routing

TylerZeroMaster committed:
Small corrections to portal guard and ternary

This should work better overall, avoiding need to check fo undefined

staxly[bot] committed:
Merge branch 'main' into core-902-content-portal-routing

staxly[bot] committed:
Merge branch 'main' into core-902-content-portal-routing

staxly[bot] committed:
Merge branch 'main' into core-902-content-portal-routing
